lib/handler/mruby/middleware.c

```diff
@@ -0,0 +1,961 @@
+        goto Failed; \
+    } \
+} while (0)
+    CHECK_REQUIRED("REQUEST_METHOD", method, 1);
+    CHECK_REQUIRED("rack.url_scheme", scheme, 1);
+    CHECK_REQUIRED("SCRIPT_NAME", script_name, 0);
+    CHECK_REQUIRED("PATH_INFO", path_info, 0);
+    CHECK_REQUIRED("QUERY_STRING", query_string, 0);
+#undef CHECK_REQUIRED
+
+    if (RSTRING_LEN(script_name) != 0 && RSTRING_PTR(script_name)[0] != '/') {
+        mrb->exc = mrb_obj_ptr(mrb_exc_new_str_lit(mrb, E_RUNTIME_ERROR, "SCRIPT_NAME must start with `/`"));
+        goto Failed;
+    }
+    if (RSTRING_LEN(path_info) != 0 && RSTRING_PTR(path_info)[0] != '/') {
+        mrb->exc = mrb_obj_ptr(mrb_exc_new_str_lit(mrb, E_RUNTIME_ERROR, "PATH_INFO must start with `/`"));
+        goto Failed;
+    }
+    if (mrb_nil_p(http_host) && (mrb_nil_p(server_name) || mrb_nil_p(server_port))) {
+        mrb->exc = mrb_obj_ptr(mrb_exc_new_str_lit(mrb, E_RUNTIME_ERROR, "HTTP_HOST or (SERVER_NAME and SERVER_PORT) is required"));
+        goto Failed;
+    }
+
+    if (!is_reprocess) {
+        /* ensure that SCRIPT_NAME is not modified */
+        h2o_iovec_t confpath = ctx->handler->pathconf->path;
+        size_t confpath_len_wo_slash = confpath.base[confpath.len - 1] == '/' ? confpath.len - 1 : confpath.len;
+        if (!(RSTRING_LEN(script_name) == confpath_len_wo_slash && memcmp(RSTRING_PTR(script_name), confpath.base, confpath_len_wo_slash) == 0)) {
+            mrb->exc = mrb_obj_ptr(mrb_exc_new_str_lit(mrb, E_RUNTIME_ERROR, "can't modify `SCRIPT_NAME` with `H2O.next`. Is `H2O.reprocess` what you want?"));
+            goto Failed;
+        }
+    }
+
+#define STR_TO_IOVEC(val) h2o_iovec_init(RSTRING_PTR(val), RSTRING_LEN(val))
+
+    /* construct url and parse */
+    h2o_iovec_t *url_comps = alloca(sizeof(*url_comps) * 9);
+    int num_comps = 0;
+    url_comps[num_comps++] = STR_TO_IOVEC(scheme);
+    url_comps[num_comps++] = h2o_iovec_init(H2O_STRLIT("://"));
+    if (!mrb_nil_p(http_host)) {
+        url_comps[num_comps++] = STR_TO_IOVEC(http_host);
+    } else {
+        url_comps[num_comps++] = STR_TO_IOVEC(server_name);
+        url_comps[num_comps++] = h2o_iovec_init(H2O_STRLIT(":"));
+        url_comps[num_comps++] = STR_TO_IOVEC(server_port);
+    }
+    url_comps[num_comps++] = STR_TO_IOVEC(script_name);
+    url_comps[num_comps++] = STR_TO_IOVEC(path_info);
+    if (RSTRING_LEN(query_string) != 0) {
+        url_comps[num_comps++] = h2o_iovec_init(H2O_STRLIT("?"));
+        url_comps[num_comps++] = STR_TO_IOVEC(query_string);
+    }
+    h2o_iovec_t url_str = h2o_concat_list(&super->pool, url_comps, num_comps);
+    h2o_url_t url_parsed;
+    if (h2o_url_parse(url_str.base, url_str.len, &url_parsed) != 0) {
+        /* TODO is there any other way to show better error message? */
+        mrb->exc = mrb_obj_ptr(mrb_exc_new_str_lit(mrb, E_ARGUMENT_ERROR, "env variable contains invalid values"));
+        goto Failed;
+    }
```

Comment:
Could you please elaborate why you need to construct `url_str` then call `h2o_url_parse` here?

Can't you just assign `super->input.scheme`, `super->input.authority`, `super->input.path` using the mruby variables e.g. `scheme`?

Reply:
This is for validation. By h2o_url_parse we validate scheme, host, port and path